### PR TITLE
add status subresources for volumeSnapshot

### PR DESCRIPTION
**Pull request by zhucan**
> /kind bug
> /kind failing-test

**Does this PR introduce a user-facing change?**:
```release-note
Add status subresource for volumeSnapshot.
```

**What this PR does / why we need it**:
add status subresources for volumeSnapshot so when update status of volumeSnapshot , it will ignore the spec filed of volumeSnapshot

**Which issue(s) this PR fixes**:
Fixes #52

### CODE REVIEW

File: cmd/csi-snapshotter/create_crd.go [hunk 1]
```diff
@@ -38,13 +38,24 @@ func CreateCRD(clientset apiextensionsclient.Interface) error {
...
 	}
-	res, err := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
 
-	if err != nil && !apierrors.IsAlreadyExists(err) {
-		klog.Fatalf("failed to create VolumeSnapshotResource: %#v, err: %#v",
-			res, err)
+	_, err := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
+	if err == nil {
+		if res, err := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Update(crd); err != nil {
+			klog.Fatalf("failed to update VolumeSnapshotResource: %#v, err: %#v",
```
Comment (xing-yang) [inline cmd/csi-snapshotter/create_crd.go]: Start external-snapshotter for the first time.  It starts without errors.  Stop it and start it again.  It failed here with the following error:

F0508 02:46:44.816579       1 create_crd.go:50] failed to update VolumeSnapshotResource: &v1beta1.CustomResourceDefinition{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1beta1.CustomResourceDefinitionSpec{Group:"", Version:"", Names:v1beta1.CustomResourceDefinitionNames{Plural:"", Singular:"", ShortNames:[]string(nil), Kind:"", ListKind:"", Categories:[]string(nil)}, Scope:"", Validation:(*v1beta1.CustomResourceValidation)(nil), Subresources:(*v1beta1.CustomResourceSubresources)(nil), Versions:[]v1beta1.CustomResourceDefinitionVersion(nil), AdditionalPrinterColumns:[]v1beta1.CustomResourceColumnDefinition(nil), Conversion:(*v1beta1.CustomResourceConversion)(nil)}, Status:v1beta1.CustomResourceDefinitionStatus{Conditions:[]v1beta1.CustomResourceDefinitionCondition(nil), AcceptedNames:v1beta1.CustomResourceDefinitionNames{Plural:"", Singular:"", ShortNames:[]string(nil), Kind:"", ListKind:"", Categories:[]string(nil)}, StoredVersions:[]string(nil)}}, err: &errors.StatusError{ErrStatus:v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:""}, Status:"Failure", Message:"customresourcedefinitions.apiextensions.k8s.io \"volumesnapshotclasses.snapshot.storage.k8s.io\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update", Reason:"Invalid", Details:(*v1.StatusDetails)(0xc0000a8360), Code:422}}

**Comment by xing-yang** [comment]
This PR does not work.  I tried to create a snapshot but it failed with the following error:

E0508 03:10:03.696019       1 goroutinemap.go:150] Operation for "create-default/new-snapshot-demo[c616fd63-713e-11e9-86c5-000c29e70439]" failed. No retries permitted until 2019-05-08 03:10:04.19600577 +0000 UTC m=+151.703720599 (durationBeforeRetry 500ms). Error: "snapshot controller failed to update default/new-snapshot-demo on API server: the server could not find the requested resource (put volumesnapshots.snapshot.storage.k8s.io new-snapshot-demo)"

File: cmd/csi-snapshotter/create_crd.go (hunk 1 above)
Comment (zhucan) [reply cmd/csi-snapshotter/create_crd.go]: I will check it again

File: pkg/controller/snapshot_controller.go [hunk 2]
```diff
@@ -831,7 +831,7 @@ func (ctrl *csiSnapshotController) updateSnapshotStatus(snapshot *crdv1.VolumeSn
 			status.RestoreSize = resource.NewQuantity(size, resource.BinarySI)
 		}
 		snapshotClone.Status = status
-		newSnapshotObj, err := ctrl.clientset.VolumesnapshotV1alpha1().VolumeSnapshots(snapshotClone.Namespace).UpdateStatus(snapshotClone)
+		newSnapshotObj, err := ctrl.clientset.VolumesnapshotV1alpha1().VolumeSnapshots(snapshotClone.Namespace).Update(snapshotClone)
```
Comment (xing-yang) [inline pkg/controller/snapshot_controller.go]: UpdateStatus was called at two places in the original PR.

Comment (zhucan) [reply pkg/controller/snapshot_controller.go]: when using "UpdateStatus" ,  the error: snapshot_controller.go:649] failed to update snapshot snapshotting-8419/snapshot-dnvgd creation timestamp: snapshot controller failed to update snapshotting-8419/snapshot-dnvgd on API server: the server could not find the requested resource (put volumesnapshots.snapshot.storage.k8s.io snapshot-dnvgd)

Comment (xing-yang) [reply pkg/controller/snapshot_controller.go]: This should not be changed to Update().  It should use UpdateStatus().  I think you need to add the following in volumesnapshot.yaml file for subresource status update to work.  See this PR: https://github.com/kubernetes/kubernetes/pull/60021/files

  subresources:
    status: {}

**Comment by xing-yang** [comment]
Please add the following to the problem description on top:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

File: pkg/controller/snapshot_controller.go (hunk 2 above)
Comment (zhucan) [reply pkg/controller/snapshot_controller.go]: ok, I will try it.

**Comment by xing-yang** [comment]
Release note needs to be in the following format:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Comment by xing-yang** [comment]
Tested creating a snapshot using example yaml files in the repo, but got the following error in the snapshot controller:

```
E0612 17:58:00.571119       1 snapshot_controller.go:370] createSnapshot [create-default/new-snapshot-demo[9e08154a-8d3b-11e9-b0aa-000c29e70439]]: error occurred in createSnapshotOperation: snapshot controller failed to update default/new-snapshot-demo on API server: the server could not find the requested resource (put volumesnapshots.snapshot.storage.k8s.io new-snapshot-demo)
E0612 17:58:00.571176       1 goroutinemap.go:150] Operation for "create-default/new-snapshot-demo[9e08154a-8d3b-11e9-b0aa-000c29e70439]" failed. No retries permitted until 2019-06-12 17:58:01.071150315 +0000 UTC m=+59.357142407 (durationBeforeRetry 500ms). Error: "snapshot controller failed to update default/new-snapshot-demo on API server: the server could not find the requested resource (put volumesnapshots.snapshot.storage.k8s.io new-snapshot-demo)"
I0612 17:58:02.876452       1 snapshot_controller.go:615] createSnapshot: Creating snapshot default/new-snapshot-demo through the plugin ...
E0612 17:58:02.893121       1 snapshot_controller.go:370] createSnapshot [create-default/new-snapshot-demo[9e08154a-8d3b-11e9-b0aa-000c29e70439]]: error occurred in createSnapshotOperation: snapshot controller failed to update default/new-snapshot-demo on API server: the server could not find the requested resource (put volumesnapshots.snapshot.storage.k8s.io new-snapshot-demo)
E0612 17:58:02.893212       1 goroutinemap.go:150] Operation for "create-default/new-snapshot-demo[9e08154a-8d3b-11e9-b0aa-000c29e70439]" failed. No retries permitted until 2019-06-12 17:58:03.893184781 +0000 UTC m=+62.179176863 (durationBeforeRetry 1s). Error: "snapshot controller failed to update default/new-snapshot-demo on API server: the server could not find the requested resource (put volumesnapshots.snapshot.storage.k8s.io new-snapshot-demo)"
```

**Comment by xing-yang** [comment]
Never mind.  Tested again on a clean setup and it worked for me.

**Pull request merged**